File: Construct Tree from Inorder & Preorder - GFG/construct-tree-from-inorder-preorder.java
// { Driver Code Starts
import java.util.*;
import java.io.*;
class Node
{
    int data; 
    Node left, right;
    Node(int key)
    {
        data = key;
        left = right = null;
    }
}

class GFG
{
    public static void main (String[] args)
    {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t-- > 0)
        {
            int n = sc.nextInt();
            Node root = null;
            int inorder[] = new int[n];
            int preorder[] = new int[n];
            for(int i = 0; i < n; i++)
              inorder[i] = sc.nextInt();
              
            for(int i = 0; i < n; i++)
              preorder[i] = sc.nextInt();
              
            Solution ob = new Solution();
            root = ob.buildTree(inorder, preorder, n);
            postOrdrer(root);
            System.out.println();
        }
    }
    
    public static void postOrdrer(Node root)
    {
        if(root == null)
          return;
          
        postOrdrer(root.left);
        postOrdrer(root.right);
        System.out.print(root.data + " ");
    }
}// } Driver Code Ends


class Solution
{
    static Node construct(int preorder[], int preStart, int preEnd, int inorder[], int inStart, int inEnd,
            HashMap<Integer, Integer> map) {
        if (preStart > preEnd || inStart > inEnd)
            return null;

        Node root = new Node(preorder[preStart]);
        int mapRootLocation = map.get(preorder[preStart]);
        int numberOfChilderen = mapRootLocation - inStart;

        root.left = construct(preorder, preStart + 1, preStart + numberOfChilderen, inorder, inStart,
                mapRootLocation - 1, map);
        root.right = construct(preorder, preStart + numberOfChilderen + 1, preEnd, inorder, mapRootLocation + 1, inEnd,
                map);

        return root;
    }

    public static Node buildTree(int inorder[], int preorder[], int n) {
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < inorder.length; i++)
            map.put(inorder[i], i);

        return construct(preorder, 0, n - 1, inorder, 0, n - 1, map);
    }
}

File: Construct Tree from Inorder & Preorder - GFG/construct-tree-from-inorder-preorder.cpp
// { Driver Code Starts
//

#include<bits/stdc++.h>
using namespace std;

struct Node
{
	int data;
	struct Node *left;
	struct Node *right;
	
	Node(int x){
	    data = x;
	    left = NULL;
	    right = NULL;
	}
};


void printPostOrder(Node *root)
{
	if(root==NULL)
		return;
	printPostOrder(root->left);
	printPostOrder(root->right);
	cout<<root->data<<" ";
}

 // } Driver Code Ends


/*Complete the code here.
Node is as follows:
struct Node
{
  int data;
  Node* left;
  Node* right;
};
*/
class Solution{
    public:
    Node* buildTree(int in[],int pre[], int n)
    {
        map<int,int>mp;
        for(int i=0;i<n;i++)
        {
            mp[in[i]]=i;
        }
        
        Node* root=tree(pre,0,n-1,in,0,n-1,mp);
        return root;
        // Code here
    }
    
    Node* tree(int pre[],int prestart,int preend,int in[],int instart,int inend,map<int,int>&mp)
    {
        if(prestart>preend||instart>inend)
        {return NULL;}
        Node* root=new Node(pre[prestart]);
        int inroot=mp[root->data];
        int numsleft=inroot-instart;
        
        root->left=tree(pre,prestart+1,prestart+numsleft+1,in,instart,inroot-1,mp);
        root->right=tree(pre,prestart+numsleft+1,preend,in,inroot+1,inend,mp);
        return root;
    }
};

// { Driver Code Starts.
int main()
{
	int t;
	cin>>t;
	while(t--)
	{
		int n;
		cin>>n;
		
		int inorder[n], preorder[n];
		for(int i=0; i<n; i++)
			cin>> inorder[i];
		for(int i=0; i<n; i++)
			cin>> preorder[i];
		Solution obj;
		Node *root = obj.buildTree(inorder, preorder, n);
		printPostOrder(root);
		cout<< endl;
	}
}
  // } Driver Code Ends